lib/rules/jsx-newline-block.js

```diff
@@ -24,18 +36,121 @@ module.exports = {
         jsxElementParents.forEach(parent => {
           parent.children.forEach((element, index, elements) => {
 
-            // JSXText here is the line breaks only
             if(['JSXElement', 'JSXExpressionContainer', 'JSXText'].includes(element.type)) {
               const prev = elements[index - 1]
               const next = elements[index + 1]
 
               if(!prev || !next) {
-                if(!prev && isEmptyLine(element)) {
-                  context.report({messageId: 'redundant', node: element})
+                // JSXText as first child
+                if(!prev && isTextElement(element)) {
```

Comment:
@quesabe do you think we can avoid nested if blocks? They can really make the business logic flow hard to perceive.

Reply:
The problem is that we have a lot of corner cases, while the correct markup stays beyond all these conditions.
I can try to get rid of the nesting as much as possible, however this might create bulky conditions. So far there are three predicates at most. With restricted nesting we might end up with a lot more, which is also not easily readable. Thus it is going to be a compromise as well - just need to chose which one is more appropriate for us (nested IFs vs condition with a lot of predicates).

Reply:
@quesabe yeah, that's clear that there loads of conditions. My point is that we can make the conditions flat with some level of redundancy, e.g.:
```typescript
if(A) {
  ...
}

if(A && B) {
  ...
}

if(A && !B && C) {
  ...
}
```

So at the cost of some repetition we'll get o the flat conditions state.

Reply:
Let me know if it makes sense and is achievable.

Reply:
So, I've done some guards, early returns and regrouping. In addition added several functions that combine a final check and reporting. Ended up with one level of nesting.

Also thought about making a single generic checking function instead of six separate. However readability will suffer. 

Let me know if you think the current approach is OK.